contracts/src/token/erc20/extensions/flashmint.rs

```diff
@@ -0,0 +1,164 @@
+
+use alloy_primitives::{Address, U256, Bytes, B256, b256};
+use stylus_sdk::msg;
+use stylus_sdk::{
+    contract, prelude::*,
+     call::{ Call, MethodError}
+};
+use crate::token::erc20::{self, Erc20, IErc20};
+use crate::flashloan::borrower::IERC3156FlashBorrower;
+use alloy_sol_types::sol;
+
+sol! {
+
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC3156UnsupportedToken(address token);
+
+
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC3156ExceededMaxLoan(uint256 maxLoan);
+
+
+     #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC3156InvalidReceiver(address receiver);
+
+}
+
+impl MethodError for Error {
+    fn encode(self) -> alloc::vec::Vec<u8> {
+        self.into()
+    }
+}
+
+pub trait  IERC3156FlashLender {
+    type Error: Into<alloc::vec::Vec<u8>>;
+
+    fn max_flash_loan(&self , token:Address) -> U256;
+
+    fn flash_fee(&self, token:Address, amount:U256)  -> Result<U256, Self::Error> ;
+
+    fn flash_loan(
+        &mut self,
+        receiver: Address,
+        token:Address,
+        amount:U256,
+        data:Bytes
+    ) -> Result<bool, Self::Error>;
+}
+
+/// A Permit error.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    UnsupportedToken(ERC3156UnsupportedToken),
+    ExceededMaxLoan(ERC3156ExceededMaxLoan),
+    InvalidReceiver(ERC3156InvalidReceiver),
+
+    /// Error type from [`Erc20`] contract [`erc20::Error`].
+    Erc20(erc20::Error),
+
+}
+
+const RETURN_VALUE: B256 =
+    b256!("439148f0bbc682ca079e46d6e2c2f0c1e3b820f1a291b069d8882abf8cf18dd9");
+
+impl IERC3156FlashLender for Erc20 {
+    
+    type Error = Error ;
+
+    /// Returns the maximum amount of tokens that can be borrowed
+    /// from this contract in a flash loan.
+    ///
+    /// For tokens that are not supported, this function returns
+    /// `U256::MIN`.
+    ///
+    /// * `token` - The address of the ERC-20 token that will be loaned.
+    fn max_flash_loan(&self, token:Address) -> U256 {
+        if token == contract::address() {
+            return   U256::MAX - self.total_supply();
+        }
+        U256::MIN
+    }
+
+    /// Calculates the fee for a flash loan.
+    ///
+    /// The fee is a fixed percentage of the borrowed amount.
+    ///
+    /// If the token is not supported, the function returns an
+    /// `UnsupportedToken` error.
+    ///
+    /// * `token` - The address of the ERC-20 token that will be loaned.
+    /// * `amount` - The amount of tokens that will be loaned.
+    fn flash_fee(&self,token:Address, amount:U256)  -> Result<U256, Error> {
+        if token != contract::address() {
+            return Err(Error::UnsupportedToken(ERC3156UnsupportedToken { token}));
+        }
+        Ok(self._flash_fee(token , amount))
+    }
+
+    /// Executes a flash loan.
```

Comment:
Docs should go into a bit more detail on what is going on inside the function. Use [Solidity version's docs](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20FlashMint.sol#L93) as inspiration.

Also missing docs sections on errors, events, panics.